crates/astria-sequencer-relayer/src/queued_blocks.rs

```diff
@@ -0,0 +1,83 @@
+use std::{
+    collections::HashMap,
+    time::{
+        Duration,
+        Instant,
+    },
+};
+
+use base64::{
+    display::Base64Display,
+    engine::general_purpose::STANDARD,
+};
+use tracing::warn;
+
+use crate::types::SequencerBlockData;
+
+// the max time to wait for a block to finalize
+const MAX_QUEUE_TIME: Duration = Duration::from_secs(10);
+
+#[derive(Default)]
+pub(crate) struct QueuedBlocks {
```

Comment:
Commented below why I think this is problematic. My suggestion is to replace this by a binary heap like like so:
```rust
struct QueuedBlocks {
    pending: HashMap<Vec<u8>, SequencerBlockData>>,
    heap: BinaryHeap<HeapValue>,
    expiration_duration: Duration,
}

struct HeapValue {
    key: Vec<u8>,
    instant: Instant,
}

}
impl PartialOrd for HeapValue {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for HeapValue {
    fn cmp(&self, other: &Self) -> Ordering {
        self.instant.cmp(&other.instant).reverse()
    }
}
impl<K> PartialEq for HeapValue<K> {
    fn eq(&self, other: &Self) -> bool {
        self.instant.eq(&other.instant)
    }
}
impl<K> Eq for HeapValue<K> {}

impl QueuedBlocks {  
    fn discard(&mut self) {
        let deadline = Instant::now() - self.expiration_duration;
        while let Some(HeapValue { instant, .. }) = self.heap.peek() {
            if *instant > deadline { return; }
            let key = self.heap.pop().unwrap().key;
            self.hash_map.remove(&key);
        }
    }
```

I haven't checked the code and it's missing insertions etc.